src/com/maddyhome/idea/vim/extension/commentary/CommentaryExtension.java

```diff
@@ -0,0 +1,138 @@
+package com.maddyhome.idea.vim.extension.commentary;
+
+import com.intellij.codeInsight.actions.MultiCaretCodeInsightActionHandler;
+import com.intellij.codeInsight.generation.CommentByBlockCommentHandler;
+import com.intellij.codeInsight.generation.CommentByLineCommentHandler;
+import com.intellij.openapi.actionSystem.DataContext;
+import com.intellij.openapi.editor.Editor;
+import com.intellij.openapi.project.Project;
+import com.intellij.psi.PsiDocumentManager;
+import com.intellij.psi.PsiFile;
+import com.maddyhome.idea.vim.VimPlugin;
+import com.maddyhome.idea.vim.command.CommandState;
+import com.maddyhome.idea.vim.command.MappingMode;
+import com.maddyhome.idea.vim.command.SelectionType;
+import com.maddyhome.idea.vim.common.TextRange;
+import com.maddyhome.idea.vim.extension.VimExtensionHandler;
+import com.maddyhome.idea.vim.extension.VimNonDisposableExtension;
+import com.maddyhome.idea.vim.key.OperatorFunction;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import static com.maddyhome.idea.vim.extension.VimExtensionFacade.*;
+import static com.maddyhome.idea.vim.helper.StringHelper.parseKeys;
+
+/**
+ * @author dhleong
+ */
+public class CommentaryExtension extends VimNonDisposableExtension {
+
+  @NotNull
+  @Override
+  public String getName() {
+    return "commentary";
+  }
+
+  @Override
+  protected void initOnce() {
+    putExtensionHandlerMapping(MappingMode.N, parseKeys("<Plug>(CommentMotion)"), new CommentMotionHandler(), false);
+    putExtensionHandlerMapping(MappingMode.N, parseKeys("<Plug>(CommentLine)"), new CommentLineHandler(), false);
+    putExtensionHandlerMapping(MappingMode.VO, parseKeys("<Plug>(CommentMotionV)"), new CommentMotionVHandler(), false);
+
+    putKeyMapping(MappingMode.N, parseKeys("gc"), parseKeys("<Plug>(CommentMotion)"), true);
+    putKeyMapping(MappingMode.N, parseKeys("gcc"), parseKeys("<Plug>(CommentLine)"), true);
+    putKeyMapping(MappingMode.VO, parseKeys("gc"), parseKeys("<Plug>(CommentMotionV)"), true);
+  }
+
+  private static class CommentMotionHandler implements VimExtensionHandler {
+    @Override
+    public void execute(@NotNull Editor editor, @NotNull DataContext context) {
+      setOperatorFunction(new Operator());
+      executeNormal(parseKeys("g@"), editor);
+    }
+  }
+
+  private static class CommentMotionVHandler implements VimExtensionHandler {
+    @Override
+    public void execute(@NotNull Editor editor, @NotNull DataContext context) {
+      final TextRange visualRange = VimPlugin.getMark().getVisualSelectionMarks(editor);
+      if (visualRange == null) {
+        return;
+      }
+
+      // always use line-wise comments
+      if (!new Operator().apply(editor, context, SelectionType.LINE_WISE)) {
+        return;
+      }
+
+      // Leave visual mode
+      executeNormal(parseKeys("<Esc>"), editor);
```

Comment:
There is `FLAG_EXIT_VISUAL` for commands. Probably you can use it?

Reply:
It's a `VimExtensionHandler` so I'm not sure where I would set that flag; `VimSurroundExtension` does the same trick.